File: MicroTDCore/Sources/MicroTDCore/Definitions/WaveDef.swift
// WaveDef.swift
// Wave composition definitions

import Foundation

/// Enemy spawn entry in wave
public struct EnemySpawn: Codable, Equatable {
    public let enemyType: String
    public let spawnTick: Int  // relative to wave start
    
    public init(enemyType: String, spawnTick: Int) {
        self.enemyType = enemyType
        self.spawnTick = spawnTick
    }
}

/// Wave definition
public struct WaveDef: Codable, Equatable {
    public let waveIndex: Int
    public let spawns: [EnemySpawn]
    public let coinReward: Int
    public let isBossWave: Bool
    
    public init(waveIndex: Int, spawns: [EnemySpawn], coinReward: Int, isBossWave: Bool = false) {
        self.waveIndex = waveIndex
        self.spawns = spawns
        self.coinReward = coinReward
        self.isBossWave = isBossWave
    }
    
    public var totalEnemies: Int {
        return spawns.count
    }
}

/// Container for all wave definitions
public struct WaveDefinitions: Codable {
    public let waves: [WaveDef]
    
    public init(waves: [WaveDef]) {
        self.waves = waves
    }
    
    public func wave(at index: Int) -> WaveDef? {
        return waves.first { $0.waveIndex == index }
    }
    
    public var totalWaves: Int {
        return waves.count
    }
}

File: expo/src/core/definitions/WaveDef.ts
// WaveDef.ts
// Wave composition definitions

/**
 * Enemy spawn entry in wave
 */
export interface EnemySpawn {
    enemyType: string;
    spawnTick: number;  // relative to wave start
}

/**
 * Wave definition
 */
export interface WaveDef {
    waveIndex: number;
    spawns: EnemySpawn[];
    coinReward: number;
    isBossWave: boolean;
}

/**
 * Container for all wave definitions
 */
export class WaveDefinitions {
    constructor(public readonly waves: WaveDef[]) { }

    /**
     * Find wave by index
     */
    wave(index: number): WaveDef | undefined {
        return this.waves.find(w => w.waveIndex === index);
    }

    get totalWaves(): number {
        return this.waves.length;
    }
}
